lib/recursive-methods.rb

```diff
@@ -1,49 +1,134 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the integer given
+# Space complexity: O(n) because the number of times the function is called is 'n' times and placed on the stack
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError
+  elsif n == 0
+    return 1
+  else
+    return n * factorial(n-1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of string 's'
+# Space complexity: O(n) because the number of times the function reverse is called and placed on the stack
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 0
+    return s
+  else
+    temp_holding = s[0]
+    return reverse(s[1..-1]) + temp_holding
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n the length of the string 's'
+# Space complexity: O(n) because of the number of functions called and placed on the stack regardless if reversed in place.
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 0 
+    return s
+  else
+    return reverse(s[1..-1]) + s[0]
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the integer 'n'
+# Space complexity: O(n), because of the function stack everytime the function is called
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  if n <= 0
+    return 0
+  elsif n == 1
+    return 2
+  else
+    return 2 + bunny(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of the string 's'
+# Space complexity: O(n) because of the function stack every time the function is called
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  def nested_search(s,index,length)
```

Comment:
It's a little ugly to nest a method inside another method.  This isn't JavaScript.